import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CMoveService.java

```diff
@@ -0,0 +1,240 @@
+
+    CMoveTask task = new CMoveTask(as, pc, cmd, keys);
+    as.getApplicationEntity().getDevice().execute(task);
+  }
+
+  private class CMoveTask extends DimseTask {
+
+    private final Attributes keys;
+
+    private CMoveTask(Association as, PresentationContext pc,
+        Attributes cmd, Attributes keys) {
+      super(as, pc, cmd);
+
+      this.keys = keys;
+    }
+
+    @Override
+    public void run() {
+      List<String> failedInstanceUids = new ArrayList<>();
+      ICStoreSender cstoreSender = null;
+      try {
+        if (canceled) {
+          throw new CancellationException();
+        }
+        runThread = Thread.currentThread();
+
+        AetDictionary.Aet cstoreTarget = aets.getAet(cmd.getString(Tag.MoveDestination));
+        if (cstoreTarget == null) {
+          sendErrorResponse(Status.MoveDestinationUnknown,
+              "Unknown AET: " + cmd.getString(Tag.MoveDestination));
+          return;
+        }
+
+        // need to get instances belonging to series/study
+        Attributes keysCopy = new Attributes(keys);
+        keysCopy.setString(Tag.QueryRetrieveLevel, VR.CS, "IMAGE");
+        String qidoPath;
+        try {
+          qidoPath = AttributesUtil.attributesToQidoPath(keysCopy,
+              TagUtils.toHexString(Tag.TransferSyntaxUID));
+          log.info("CMove QidoPath: " + qidoPath);
+        } catch (DicomServiceException e) {
+          log.error("CMove QidoPath error");
+          sendErrorResponse(e.getStatus(), e.getMessage(), e.getDataset(), null);
+          return;
+        }
+
+        JSONArray qidoResult;
+        try {
+          qidoResult = dicomWebClient.qidoRs(qidoPath);
+          if (qidoResult == null || qidoResult.length() == 0) {
+            throw new IDicomWebClient.DicomWebException("No instances to move");
+          }
+        } catch (IDicomWebClient.DicomWebException e) {
+          log.error("Failed QidoRs for CMove", e);
+          sendErrorResponse(Status.UnableToCalculateNumberOfMatches, e.getMessage());
+          return;
+        }
+
+        cstoreSender = cstoreSenderFactory.create();
+
+        int successfullInstances = 0;
+        int remainingInstances = qidoResult.length();
+        for (Object instance : qidoResult) {
+          sendPendingResponse(remainingInstances, successfullInstances, failedInstanceUids.size());
+
+          if (canceled) {
+            throw new CancellationException();
+          }
+
+          JSONObject instanceJson = (JSONObject) instance;
+          String studyUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.StudyInstanceUID));
+          String seriesUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.SeriesInstanceUID));
+          String instanceUid = AttributesUtil.getTagValue(instanceJson,
+              TagUtils.toHexString(Tag.SOPInstanceUID));
+          String classUid = AttributesUtil.getTagValue(instanceJson,
```

Comment:
Would this cause an error on CSTORE for when TS or SOP ClassUID are missing? Alternatively we can just call getTagValue for both, since that throws an exception if the tag is not found.